### DISCUSSION TITLE
GTD workflow

### DISCUSSION THREAD

**Discussion by danymat** [nvim-neorg/neorg#61] [Ideas]
Hello everyone, I'm starting to implement the Getting things done methodology, and I think we need to discuss about the workflow we will be using for.
If you don't know what is GTD, or want a quick recap of the capabilities, you can see: https://hamberg.no/gtd/

- You can check out the current PR for the base module here: https://github.com/vhyrro/neorg/pull/60

There are some issues that need to be addressed :

## Contexts management 

How do we handle contexts (or tags)? Do we create a file for each context ? 

This can be a problem if I want to assign multiple contexts to a task. We could use something similar to todo-txt workflow: add some metadatas at the end of the task like this 
```
- [ ] this is a task @home @phone
```

This way we could afterwards present the tasks that have a specific tasks, by filtering them

## Projects management

The base idea in GTD is to have a list containing all projects.
If we follow it, we have to create a file called projects, which will only be a list of bullet points for each project title. 

The problem is that it's difficult to track the tasks related to a project during the day.

We could use the same idea than contexts, which is with tags.

The todo-txt workflow specify +projectName for a task, like this 

```

- [ ] this is a task @home @phone +thisIsAProject
```

What I don't like about it is the obligation to camel case the name, so we could think about a better idea

## Tasks/projects related notes

Some tasks or projects require notes in order to remember things or to add links/insights and ideas.
We need to have a way to store those related notes

**EDIT 12/08**

I suppose we can use the way @vhyrro did with tables, such as :
```
- [ ] This is a task
@notes
This is the note of the task
@end
```

Or for a full note file:

```
- [ ] This is a task
@notes
[note link](note link.norg)
@end
```

## And more..

I will be addressing more issues afterward (e.g how do we handle completed tasks or projects), but this is a good starting point. When updating the post I will comment so you can know what's been added

Fell free to participate in the discussion everyone !

**Comment by danymat** [comment]
Added section **task related notes**

**Comment by notusknot** [comment]
I love this idea, I'm very excited for whats to come as well. As neorg expands and I might write a blog post on how to set up a full organization workflow with neorg

**Comment by danymat** [comment]
Edited **task related notes** (see edit 12/08)

**Comment by vhyrro** [comment]
Alright, it's time that I step away from development a bit and focus on giving ideas for this :)

#### Short Disclaimer
If there's something I say that is blatantly stupid plz call me out. Some ideas may blend in with the concept of org-agenda - if that's something
you wouldn't like also tell me (although I believe it makes sense). The syntax I propose is certainly not final, however the concepts themselves should
serve as a nice foundation.

You may also see me slightly move away from the "traditional" ways of doing things. I think it's good to move away to potentially better solutions as long as it's not annoying and difficult
to learn.

# General Workflow
In my opinion some changes will need to be made in order to make the experience nice. Most importantly
the organization of notes. I really wouldn't like things to cram together after a while and look very messy.
Lemme propose how things could be done:

First of all, capturing. How do we capture new ideas? There can be a few ways to achieve such a thing. The most basic, however, is _single-line captures_.
You press a key and get a popup. Here you can enter a note. Let's say I need to create a presentation for school in 2 days, without thinking much I type this:

```
Create presentation @school +Presentation
```

And press enter. Ok, what happens now? Where does this go? All notes should be stored inside of an inbox file, as you would expect :)

The actual name of this file should be configurable - there's no need for it to be a static filepath. It should exist in the current workspace though, as almost everything should.
What I've just made is a note with the content "Create presentation" and the `school` context within the `presentation` project. You remember when I said I don't like things to cram together
and be very messy? This is how the note gets stored within the inbox file:

```
* Presentation
  ** School
     - [ ] Create presentation
```

Do we have several projects? Rather simple:

```
* Project1
  ** Category (or context or whatever you wanna call it)
     - [ ] Note!

* Other project
  - [ ] Note with no category!
```

Now I know a few issues may pop into your head right away but I want you to hold out, I'll address those later.

## Selecting a Date and Time for a Task
Tbh I dunno about the `due:<time>` stuff from todo.txt and others. It's a lot to write. Writing can be hard when you have a lot of ideas flowing in a stream.
Have you used Neogit? Are you familiar with the popups that appear there in a split, where you can press a key that corresponds to an action
(e.g. if i press `c` for commit I can then press `c` for a new commit or `a` to amend etc.)? It would be nice to have something like that.
I create a note, then I have a few options: I can either press `s` for "someday", `t` for `tomorrow`, `<count>t` for `in <count> days`, `<count>?w` for `next week`/`in <count> weeks` (if the count was provided) and so on.
I could also have a `c` for "custom" which would bring up a calendar popup similar to how it is in org-agenda where I can select a custom date myself.

Just like that, I don't need to think. I write a note, then out of pure muscle memory press `w` and it's scheduled for next week, convenient right? :)

# How are dates stored? + Other questions
The proposed syntax for the inbox file is:
```
* <Project Name>
  ** <Category>
     $due <date>
     - [ ] Note
     - [ ] Other note due the same day

     - [ ] Note not bound to any due date (because it's divided by a hard line break from the others)
```

Very nicely categorized and doesn't get all over the place really quickly.
Here's where you might start getting some ideas, what if I do:
```
A note @home @phone
```

Uh oh! Where would we store it now? The same note in two subheadings, like so?:
```
* General
  ** Home
     - [ ] A note
  ** Phone
     - [ ] A note
```

Of course not! Ugly, ugly, ugly. And a pain to manage. Time to use IDs and custom names!

```
* General
  $name feb4-5c2a-3fc1
  - [ ] A note

  ** Home
     -> [A note](#feb4-5c2a-3fc1)
  ** Phone
     -> [A note](#feb4-5c2a-3fc1)
```

We could always use numbers instead of hashes or whatever you please. The idea is that notes will simply become links to a single note that exists in the top level of the project.
If the note spans several projects then the links will be placed at the top-level of the document:

```
$name feb4-5c2a-3fc1
- [ ] A note

* Project1
  ** Category1
     -> [A note](#feb4-5c2a-3fc1)

* Project2
  -> [A note](#feb4-5c2a-3fc1)
```

Ok so right now this seems like one of the biggest pains in the ass to parse - or so you would think! With the slow rise of the new treesitter parser I'm working on parsing *everything* I proposed
is actually incredibly easy. Every single element can be easily distinguished with no extra logic necessary - all you have to do is look at where the nodes are in the syntax tree. Don't worry, I'll definitely
be the one handling that lol

I wanna go to these great lengths of having things laid out because I hate having an organizational tool that does no organization internally :P

# How the Inbox is Viewed
With all the nice stuff we're doing you could already be on your merry way, but I'm not done yet. Realistically you shouldn't be viewing the inbox raw, i.e. you shouldn't be manually modifying it.

What I would like is a view of the inbox inspired by the org-agenda view.
If you haven't seen it, this is what it looks like (we're not gonna be copying it exactly, I just want you to get a decent idea):

![agenda](https://user-images.githubusercontent.com/76052559/129412990-e4d3bf2c-7f78-462c-b79b-2e9869675ff1.png)


You will be able to see a list of things you need to get done for a certain date, as well as have a bunch of functionality exposed in order
to fulfill the other steps of GTD (like filtering tasks that don't have a schedule etc.).


# Soo much to talk about
I've barely scratched the surface, but that's all I really have time for today, sorry! I've talked a lot about the internals (which are important when starting out) and I hope to at least spark
some potential future ideas. Remember, what I say isn't set in stone, we need feedback to become as good as we can!

Here are the other things I could potentially talk about + much more:
- Different ways of capturing notes and tasks
- Custom keybinds in `inbox-mode` (there can be a lot of them)
    - Manipulating tasks from within the inbox via key combos
- How projects can be manipulated
- Scheduling for a specific time *and* date (rather than just date)
- Filtering unscheduled tasks
- Automatic scheduling of tasks
    - Sometimes we as humans don't like thinking and making choices, that's why we should make a machine think for us! We can tell Neorg to simply assign a schedule to a task wherever it sees fit:
        - Being able to set working hours so Neorg can know when it should autoassign certain tasks
        - Allowing neorg to automatically assign different types of tasks for different times of day (for example we may want to have our home related tasks be assigned after our working hours somewhere closer to the evening)
        - Making Neorg be able to contextually shove certain small scale tasks inbetween larger tasks that are spread apart

# That's it!
That's all I've got time for today - let me know, do you think the ideas are good? Do you think they're hot garbage? Do you think they're a great starting point? Does it completely derange from the GTD philosophy and should be thrown in the trash?
Haha, I'm excited to see what you guys think of at least these basic concepts.
Also if the stuff I wrote feels a bit rambly/incoherent it's because it's late and my brain is rather fried. In the meantime feel free to ask questions if something I wrote doesn't make sense.

**Reply by vhyrro** [reply]
I have that weird feeling I missed something blatantly obvious and I'm gonna be ridiculed. It might just be late though and I might be paranoid lmfao

**Reply by danymat** [reply]
Hello everyone, and specially @vhyrro. I will focus my reply on asking more precisions and approving the parts that I'm okay with.
*Sidenote: I will write a brief summary of the things discussed in Discord group for audience purpose.*

## Creating a task, syntax

> `Create presentation @school +Presentation`

For capturing notes I will use this syntax, and I propose to specify the syntax a little bit more:

- Every text before an `@`, a `+` or a `$due:`, `$start`, `$note:` will be treated as `Task`
- After a `Task`, I can specify all or none of the optional fields (`@`,`+`, etc)
- The word following directly the `@` is considered as a context (tag)
- Project
  - The word following directly the `+` is considered as a Project 
  - If I want a project name to be multiple words, I will specify double quotes after the `+` like this: `+"This is a project"`
- When I want a task to be shown after a certain starting date, I specify `$start:`, like this: `$start:2021-03-21`or `$start:2w`(2 weeks from now)
- When I want a task to be due for a certain date, I specify `$due:`, like this: `$due:2021-03-21`or `$due:2w`(2 weeks from now)
- For notes, we can use the same concept as project names with spaces: `$note:""`, like this: `$note:"This is a note related to the task"`
- I can put multiple contextes like this `@context1 @context2`, but every other fields must appear 0 or 1 times.
 
## Creating a project 

I could be able to create a project without creating a task in it, so the syntax is quite the same as creating a task, but:

- I must start the capturing with `+""` or `+`
- Every other field is allowed after the project name (`@`, a `+` or a `$due:`, `$start`, `$note:`)
  - That means the fields written for the project are specific for the project and for all tasks within it
 
## Examples

```
Create presentation @school +"French class" @due:tomorrow
```

will be treated as

```
* French Class
  ** school
    - [ ] Create presentation $due:2021-08-22
```

Why did I not put the `$due`field before the task ? I think it could mess the interpretation if there are multiple tasks one after the other, because I would not know if the due is for all tasks or just for the task below

--- 

```
+"Winter Vacations - Preparation" $start:2021-12-20 @home
```

will be treated as (note the context location, I'm not sure this is the best idea though for context placement, we need to discuss on that. I used $start after the name the same way I did with the `$due:` in the example above

```
* Winter Vacations - Preparation $start:2021-12-20 @home
```

## Summary of discord group

- For a task with multiple contexts, like 

```
Task @home @mac
```

we (I and @vhyrro) think that creating a link like this :

```
* General
  $name feb4-5c2a-3fc1
  - [ ] A task

  ** home
     -> [A task](#feb4-5c2a-3fc1)
  ** mac
     -> [A task](#feb4-5c2a-3fc1)
```

can lead to complexity in reading the file without Neorg.

We could use 

```
* General
  ** home mac 
    - [ ] A task
```

Which is cleaner, but loses a bit of structure

## Considerations

I only replied about the structure of the quick capturing, but I think after this is done, I will create a more advanced capturing in the same way you proposed:

For example, at the moment the capturing looks like this:

<img width="605" alt="Capture d’écran 2021-08-14 à 15 02 08" src="https://user-images.githubusercontent.com/5306901/129447118-17d30e10-62c4-4f21-a1a3-9b3023a9b66c.png">

- If I don't specify an optional field (so there's only a task created) and I press `<CR> `,there is a list of commands that shows up, like in your proposition :

```
Have you used Neogit? Are you familiar with the popups that appear there in a split, where you can press a key that corresponds to an action
(e.g. if i press c for commit I can then press c for a new commit or a to amend etc.)? It would be nice to have something like that.
I create a note, then I have a few options: I can either press s for "someday", t for tomorrow, <count>t for in <count> days, <count>?w for next week/in <count> weeks (if the count was provided) and so on.
I could also have a c for "custom" which would bring up a calendar popup similar to how it is in org-agenda where I can select a custom date myself.
```

To select all fields I want to add (project name, due/start date, contexts, note)

**Reply by mrossinek** [reply]
I think there is a lot of useful stuff in here which is definitely a good starting point. However, seeing where this discussion is going, I shortly want to raise some (minor ?) concerns:

### Multiple Files
How will Neorg be able to handle tasks being split across multiple files?
Consider a scenario where you work on completely Independent projects and want to separate your notes for those by creating separate `.norg` files. If you now want to store the tasks related to each project in that respective file, how will Neorg be able to e.g. generate a `next` report?
It would have to know about/parse all `.norg` files at once, right? I think, within a single workspace that may be feasible (please correct me if I'm wrong) but once you want to separate e.g. personal and work projects into different workspaces, you also need to deal with two separate task reports.

### Workspaces
On that topic, how do workspaces integrate into this? Personally, I am not sure yet how/if I will be using those exactly for the concern I wrote above. This means I would probably rather keep all my personal and work related tasks inside a single workspace.
Another issue with regards to workspaces is the inbox: when I quickly note down a task, where does it go? Presumably the inbox of the current workspace. What if I am at a work and a personal task comes up: Can I move that easily to another workspace? Or is there a workspace-independent inbox?

---

### Side note: Taskwarrior
I remember reading somewhere that @vhyrro wrote that taskwarrior support may come at any point. I am a big fan of that and I think that having a centralized database for tasks (like provided by taskwarrior) would likely resolve most of my concerns above.
However, we obviously don't want to depend on taskwarrior and provide a full workflow with pure Neorg. I just bring this up to potentially keep this in mind while planning some things, because looking at taskwarrior early may help us with respect to some design decisions.

---

### Syntax
Just a quick question: is using `+` as a prefix for project used commonly? In taskwarrior I am used to using `+` for tags (and I tend to write context-related tags as e.g. `+@home`). I am just asking because it threw me off while reading the above.

**Reply by danymat** [reply]
> Consider a scenario where you work on completely Independent projects and want to separate your notes for those by creating separate .norg files. If you now want to store the tasks related to each project in that respective file, how will Neorg be able to e.g. generate a next report?

Yes, fetching all Neorg files in the gtd workspace will create the reports from it. 

> On that topic, how do workspaces integrate into this? Personally, I am not sure yet how/if I will be using those exactly for the concern I wrote above. This means I would probably rather keep all my personal and work related tasks inside a single workspace.

Atm on #60, there is only one and only one workspace for all gtd purposes. In this manner, a workspace is like an "Application" like Omnifocus or things. 
In order to distinguish work and personal tasks (what's called horizons of focus in gtd: https://gettingthingsdone.com/2011/01/the-6-horizons-of-focus/), I have an idea: Creating folders for each horizon of focus, and in there the respective Neorg files.

> Another issue with regards to workspaces is the inbox: when I quickly note down a task, where does it go? Presumably the inbox of the current workspace. What if I am at a work and a personal task comes up: Can I move that easily to another workspace? Or is there a workspace-independent inbox?

Atm on PR #60 , you have to specify the workspace for gtd like this :
```
load = {
        ["core.defaults"] = {},
        ["core.keybinds"] = {
            config = {
                default_keybinds = true,
                neorg_leader = "<Leader>o"
            }
        },
        ["core.norg.concealer"] = {}, -- Allows for use of icons
        ["core.norg.dirman"] = { -- Manage your directories with Neorg
            config = {
                workspaces = {
                    main = "~/neorg",
                    gtd = "~/gtd"
                }
            }
        },
        ["core.gtd.base"] = {
            config = { workspace = "gtd" }
        },
        ["core.ui"] = {},
    },
```

The inbox file is called inbox.neorg and placed on this workspace on root folder. (the inbox name will be customizable)
So with my idea that involves a single workspace, if you are at work and want to move the task, you certainly have to specify the Neorg file to move. 

With respect to what I replied, a simple file structure could be:

```
.
├── inbox.norg
├── personal
│   └── all.norg
└── work
    ├── methodologies.norg
    └── reunions.norg
```

> Just a quick question: is using + as a prefix for project used commonly? In taskwarrior I am used to using + for tags (and I tend to write context-related tags as e.g. +@home). I am just asking because it threw me off while reading the above.

~I think that it is used in todoist too~ Nervermind they use `#`

**Reply by mrossinek** [reply]
> Atm on #60, there is only one and only one workspace for all gtd purposes.

I see. This kind of circumvents potential issues arising from multiple workspaces. I would be fine with such a limitation but I guess @vhyrro will have a word or two so say about this.

> Nervermind they use #

Personally, I would prefer that. However, I think we should attempt to unify the syntax with the rest of Neorg's philosophy as much as possible. Again @vhyrro probably has the best overview of which modifiers to use here.

**Reply by vhyrro** [reply]
I think having a separate workspace for GTD sounds cool to be fair. If you're gonna get things done you might as well hyperfocus. This doesn't mean we should make it difficult for others to quickly jump around though. When adding a quick task from another workspace you'd still want that task to end up in the GTD workspace, right? In that scenario the expected behaviour would be to set the workspace to the GTD one, drop the task where we need and then return to the other workspace without interruptions. We could even add some extra functions to the API that will allow us to view and modify other workspaces without the need of switching to them. Opening the inbox though would require us to physically switch to the GTD workspace though, as that's more serious than just quickly dropping a task.

As for syntax, correct me if I'm wrong but `#` sounds more fitting for a context, doesn't it? Like, if you look at how people are used to doing things in social media, `#` is for referencing a hashtag whilst `@` is used for referencing a user. Would it make sense to have `#` for contexts and `@` for projects 🤔? Just throwing out some ideas. I have some other ideas regarding syntax but I'll have to play around with them a bit more in order to make sure they make sense in the context of Neorg. I'll keep y'all posted.

**Reply by danymat** [reply]
I don't have preferences regarding # for projects, but I think an `@` for context is relevant because it is commonly read `at`, like `@home` can be read `at home`

**Reply by mrossinek** [reply]
I agree that tasks should always go into the GTD workspace (if we want a single one). And these operations should be as frictionless as possible :+1: I see that in #60 you currently configure the GTD workspace by name. Assuming the inbox is called `inbox.norg` at the root of that workspace, opening that file in (e.g.) a split should be pretty straight forward, right? You could then switch the workspace for that buffer only (if dirman supports per-buffer/per-window workspaces (see `:lcd`)).

Since I am only starting out with GTD as a proper workflow, I am not 100% sure about contexts. I come from a "do-whatever-works" approach with taskwarrior. They only have a concept of `projects` and `tags` (on purpose) where you would use tags to interpret anything like contexts, etc.
Projects are set via the `project:` keyword (just like `due:`, `wait:` etc.) but I would strongly like a special syntax modifier for this, because it is so common. Tags are added with `+` and removed with `-`. In the past, I would use tags like `+@home`, `+@work` to set (what I called) a context. Reading `at home` simply made sense to me.
Thinking of `#` in terms of github, where we use it to refer to issues or PRs, I think it seems natural to associate that with a project (which can seen as an issue or a PR), right?

**Comment by danymat** [comment]
Current state of #60 is very promising, with a full quick inbox entry working with support for a full range of syntax!

@vhyrro just released a new killer featured called `selection_popup` that can allow us to create a small view asking us for selection.
It can even nest selections. Here is a demo of it:

https://user-images.githubusercontent.com/5306901/129937978-a28a837c-37cf-4f4a-a3f1-7abd7618e889.mov

We now need to think about the use cases we could use it. 
My first guess:

- Doing operations on a task (pressing a command on a task will allow it to open), like move a task, archive it, add due dates, add note, add context, etc

Feel free to participate and explain your choices!

**Comment by danymat** [comment]
Hello everyone, I created a documented detailing the full specs of GTD in neorg: #67
